### DISCUSSION TITLE
Commercial vs Commercial Plus - What are the details? Can we upgrade later if needed?

### DISCUSSION THREAD

**Discussion by nrmiller** [PrismLibrary/Prism#3271] [General] [answered]
Hello! We are planning to use Prism for our .NET MAUI project. I was hoping to know if there were any additional details on the included features.

Specifically, Commercial Plus has:

1. Additional Containers
2. Install via Private NuGet feed.

What are the additional containers? Is DryIoc available in the Commercial edition, or would Commercial Plus be needed?
Is there a reason to use the private feed other than accessing more frequent builds/updates?

**For the most part, we think the commercial edition would be fine, but if we choose incorrect, can we later upgrade to Commercial Plus at reduced cost?**

I've only seen the comparison at the [pricing page](https://prismlibrary.com/#pricing), but any additional details would be appreciated.

Thank you!

**Comment by dansiegel** [comment, marked as answer]
> we think the commercial edition would be fine

Honestly we've found that the majority of people are opting for the Commercial Plus because we have made every effort to provide a lot of extra value, but we encourage you to make the choice that is right for you and your team.

> if we choose incorrect, can we later upgrade to Commercial Plus at reduced cost?

Short answer is yes you can upgrade paying the difference between the licenses.

> Is DryIoc available

It is available to users with all licenses, including those who qualify for the Community License.

> What are the additional containers?

DryIoc is a fantastic container and continues to be my container of choice, and I continue to recommend that anyone capable of it, please consider sponsoring [@dadhi](https://github.com/sponsors/dadhi). That said we recognize that having a single DI Container supported poses a security risk for the Prism Community. As a result we have taken the time to provide support for some additional containers to help safeguard your software investments.

Currently, Commercial Plus users have access to two additional Containers
- [Grace](https://github.com/ipjohnson/Grace)
- Microsoft.Extensions.DependencyInjection - note that this uses a private fork from .NET 8 to allow us to make the container mutable

Note that we may add additional containers in the future. For instance we do have an open request for Castle Windsor.

> Is there a reason to use the private feed other than accessing more frequent builds/updates? 
> (benefits to the Commercial Plus License)

In addition to the private NuGet feed you also get access to the Prism Discord server where you can ask questions. We find the Prism Community to be a fantastic one full of individuals who step up to help out. @brianlagunas and I try to answer questions as we can, but often times you may also find that another community member has encountered the same issue before and they will step up to help you out far faster than other public methods including GitHub Discussions. Do keep in mind that the Discord is not a substitution and should not be confused with actual Enterprise Support. 

As for the private NuGet feed, in addition to having access to bug fixes/features for Prism as soon as they're merged you also have access to the Prism Plugins. 

**Plugin Popups**
For those looking to get the benefit of the [Mopups](https://github.com/LuckyDucko/Mopups) with their Prism application for Dialogs you can use the [Popups Plugin](https://docs.prismlibrary.com/docs/plugins/popups.html). Note that if you've used this in the past for Xamarin.Forms that we no longer support Navigation with Popups as a Popup is more of a Dialog and should be used with the Prism Dialogs API.

**Plugin Logging**
Our [Logging Plugin](https://docs.prismlibrary.com/docs/plugins/logging/index.html) which has a pretty robust set of Providers including AppCenter, Raygun and Sentry. This can be highly customized to allow you to send Errors to one provider and Analytics to another, or A/B test and evaluate what platform works best for your needs.

I am also currently working on putting together a demo that makes use of Aspire to launch Raygun locally with Ollama, Open WebUI and Graylog Server. This will give you a complete dev setup to capture logs, events, and errors with AI integration from the Raygun Team to analyze your errors, and Open WebUI to give you a private/local portal to interrogate the model(s) of your choice on anything you may want without concerns of your private data being used by any 3rd parties.

**Plugin Essentials**
The [Essentials Plugin](https://docs.prismlibrary.com/docs/plugins/essentials/index.html?tabs=maui) is a bit of a cross between the features of Xamarin/MAUI Essentials and some of what's available in Shiny. This library is UI Framework agnostic giving you greater power to keep your code independent from being tightly coupled with a UI Framework. This means that when using Region Navigation and Prism.Plugin.Essentials you can create a Service Layer complete with ViewModels, Notifications, etc and reuse that code on a WPF, .NET MAUI or Uno Platform application today and in the future we expect to extend that to Avalonia once #3127 is merged.

One of my personal favorites in the Essentials Plugin is that this is 100% of the code you need to write to implement a Secure Store and access your secure value in code. As you'll see though in the docs there are a lot more APIs either already available or that will be available in the coming weeks.

```cs
[SecureStore]
public partial interface IUserStore
{
    string? Jwt { get; set; }
}

container.RegisterStore<IUserStore>();
```

**Plugin Observable Regions**
Regardless of which UI Framework you choose, if you're developing an application with Region Navigation you can take advantage of the [Prism.Plugin.ObservableRegions](https://docs.prismlibrary.com/docs/plugins/regions.html?tabs=maui) package. This provides an IObservable that allows you to globally listen to and respond to Navigation events within all of your applications Regions. This means that rather that needing to wire up handlers all throughout your code to say provide an error Dialog or log the exception, you can do this in a single place in your application.

## General Note

Many of the features within Essentials and other plugins/tooling that we provide has also come from companies privately sponsoring these features. So if there is a feature that you require we can have the conversation with you to see what it might take to get additional features that you're looking for.
